src/signedTransaction.ts

```diff
@@ -0,0 +1,171 @@
+import { Transaction } from './transaction.js';
+import { LogicSig } from './logicsig.js';
+import {
+  EncodedMultisig,
+  encodedMultiSigToEncodingData,
+  encodedMultiSigFromEncodingData,
+  ENCODED_MULTISIG_SCHEMA,
+} from './types/transactions/index.js';
+import {
+  AddressSchema,
+  FixedLengthByteArraySchema,
+  NamedMapSchema,
+} from './encoding/schema/index.js';
+
+export class SignedTransaction implements Encodable {
+  static encodingSchema = new NamedMapSchema([
+    {
+      key: 'txn',
+      valueSchema: Transaction.encodingSchema,
+      required: true,
+      omitEmpty: true,
+    },
+    {
+      key: 'sig',
+      valueSchema: new FixedLengthByteArraySchema(64),
+      required: false,
+      omitEmpty: true,
+    },
+    {
+      key: 'msig',
+      valueSchema: ENCODED_MULTISIG_SCHEMA,
+      required: false,
+      omitEmpty: true,
+    },
+    {
+      key: 'lsig',
+      valueSchema: LogicSig.encodingSchema,
+      required: false,
+      omitEmpty: true,
+    },
+    {
+      key: 'sgnr',
+      valueSchema: new AddressSchema(),
+      required: false,
+      omitEmpty: true,
+    },
+  ]);
+
+  /**
+   * The transaction that was signed
+   */
+  public readonly txn: Transaction;
+
+  /**
+   * Transaction signature
+   */
+  public readonly sig?: Uint8Array;
+
+  /**
+   * Multisig structure
+   */
+  public readonly msig?: EncodedMultisig;
+
+  /**
+   * Logic signature
+   */
+  public readonly lsig?: LogicSig;
+
+  /**
+   * The signer, if signing with a different key than the Transaction type `sender` property indicates
+   */
+  public readonly sgnr?: Address;
+
+  constructor({
+    txn,
+    sig,
+    msig,
+    lsig,
+    sgnr,
+  }: {
+    txn: Transaction;
+    sig?: Uint8Array;
+    msig?: EncodedMultisig;
+    lsig?: LogicSig;
+    sgnr?: Address;
+  }) {
+    this.txn = txn;
+    this.sig = sig;
+    this.msig = msig;
+    this.lsig = lsig;
+    this.sgnr = sgnr;
+
+    let numberOfSigs = 0;
+    if (sig) numberOfSigs += 1;
+    if (msig) numberOfSigs += 1;
+    if (lsig) numberOfSigs += 1;
+    if (numberOfSigs > 1) {
+      throw new Error(
+        `SignedTransaction must not have more than 1 signature. Got ${numberOfSigs}`
+      );
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  public getEncodingSchema() {
+    return SignedTransaction.encodingSchema;
+  }
+
+  public toEncodingData(): Map<string, unknown> {
+    const data = new Map<string, unknown>([
+      ['txn', this.txn.toEncodingData()],
+      ['sig', this.sig],
+      ['sgnr', this.sgnr],
+    ]);
+    if (this.msig) {
+      data.set('msig', encodedMultiSigToEncodingData(this.msig));
+    }
+    if (this.lsig) {
+      data.set('lsig', this.lsig.toEncodingData());
+    }
+    return data;
+  }
+
+  public static fromEncodingData(data: unknown): SignedTransaction {
+    if (!(data instanceof Map)) {
+      throw new Error(`Invalid decoded SignedTransaction: ${data}`);
+    }
+    return new SignedTransaction({
```

Comment:
should there be a check for more than one signature on decode?

Reply:
I'm not so sure about this. If you try to create a signed txn with more than one signature, then yes we should error and stop you because you're doing something wrong and it's best to fail early.

But if you're just trying to read such a transaction that someone else created, I don't see a clear benefit to erroring in that case. There are almost an unlimited number of ways you can invalidate a transaction, and I'm not sure it's worth the effort to try to validate every field completely upon decoding (we do validate that field types are correct, but field values/consistency is another thing entirely). That will be done when you send the transaction to a node for processing anyway.